flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java

```diff
@@ -598,189 +535,154 @@ protected int cancel(String[] args) {
+			if (withSavepoint) {
+				if (targetDirectory == null) {
+					logAndSysout("Cancelling job " + jobId + " with savepoint to default savepoint directory.");
 				} else {
-					logAndSysout("Cancelling job " + jobId + '.');
-					client.cancel(jobId);
-					logAndSysout("Cancelled job " + jobId + '.');
+					logAndSysout("Cancelling job " + jobId + " with savepoint to " + targetDirectory + '.');
 				}
+				String savepointPath = client.cancelWithSavepoint(jobId, targetDirectory);
+				logAndSysout("Cancelled job " + jobId + ". Savepoint stored in " + savepointPath + '.');
+			} else {
+				logAndSysout("Cancelling job " + jobId + '.');
+				client.cancel(jobId);
+				logAndSysout("Cancelled job " + jobId + '.');
+			}
 
-				return 0;
-			} finally {
+			return 0;
+		} finally {
+			try {
 				client.shutdown();
+			} catch (Exception e) {
+				LOG.info("Could not properly shut down the client.", e);
 			}
 		}
-		catch (Throwable t) {
-			return handleError(t);
-		}
 	}
 
 	/**
 	 * Executes the SAVEPOINT action.
 	 *
 	 * @param args Command line arguments for the cancel action.
 	 */
-	protected int savepoint(String[] args) {
+	protected int savepoint(String[] args) throws CliArgsException {
 		LOG.info("Running 'savepoint' command.");
 
-		SavepointOptions options;
-		try {
-			options = CliFrontendParser.parseSavepointCommand(args);
-		} catch (CliArgsException e) {
-			return handleArgException(e);
-		} catch (Throwable t) {
-			return handleError(t);
-		}
+		SavepointOptions options = CliFrontendParser.parseSavepointCommand(args);
 
 		// evaluate help flag
 		if (options.isPrintHelp()) {
 			CliFrontendParser.printHelpForSavepoint();
 			return 0;
 		}
 
-		if (options.isDispose()) {
-			// Discard
-			return disposeSavepoint(options);
-		} else {
-			// Trigger
-			String[] cleanedArgs = options.getArgs();
-			JobID jobId;
+		CustomCommandLine<?> customCommandLine = getActiveCustomCommandLine(options.getCommandLine());
 
-			if (cleanedArgs.length >= 1) {
-				String jobIdString = cleanedArgs[0];
-				try {
-					jobId = new JobID(StringUtils.hexStringToByte(jobIdString));
-				} catch (Exception e) {
-					return handleArgException(new IllegalArgumentException(
-							"Error: The value for the Job ID is not a valid ID."));
-				}
+		ClusterClient clusterClient = customCommandLine.retrieveCluster(options.getCommandLine(), config, configurationDirectory);
+
+		try {
+			if (options.isDispose()) {
+				// Discard
+				return disposeSavepoint(clusterClient, options.getSavepointPath());
 			} else {
-				return handleArgException(new IllegalArgumentException(
-						"Error: The value for the Job ID is not a valid ID. " +
-								"Specify a Job ID to trigger a savepoint."));
-			}
+				// Trigger
+				String[] cleanedArgs = options.getArgs();
+				JobID jobId;
+
+				if (cleanedArgs.length >= 1) {
+					String jobIdString = cleanedArgs[0];
+					try {
+						jobId = new JobID(StringUtils.hexStringToByte(jobIdString));
```

Comment:
Might as well use `JobID.fromHexString()` and add a meaningful message in `fromHexString` that you propagate.